tcpdevmem_udma.c

```diff
@@ -0,0 +1,309 @@
+#define __iovec_defined 1
+
+#include <linux/memfd.h>
+#include <linux/if.h>
+#include <linux/dma-buf.h>
+#include <linux/socket.h>
+#include <linux/udmabuf.h>
+#include <linux/uio.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+
+#include <stddef.h>
+#include <fcntl.h>
+
+#include "flow.h"
+#include "lib.h"
+#include "logging.h"
+#include "tcpdevmem.h"
+#include "tcpdevmem_udma.h"
+#include "thread.h"
+
+#define TEST_PREFIX "ncdevmem_udma"
+
+int udma_setup_alloc(const struct options *opts, void **f_mbuf, struct thread *t)
+{
+        bool is_client = opts->client;
+        int devfd;
+        int memfd;
+        int buf;
+        int buf_pages;
+        int ret;
+        size_t size = opts->tcpd_phys_len;
+
+        struct tcpdevmem_udma_mbuf *tmbuf;
+        struct dma_buf_create_pages_info pages_create_info;
+        struct udmabuf_create create;
+
+        if (f_mbuf == NULL)
+                return ENOMEM;
+
+        if (*f_mbuf)
+                return 0;
+
+        tmbuf = (struct tcpdevmem_udma_mbuf *)calloc(1, sizeof(struct tcpdevmem_udma_mbuf));
+        if (!tmbuf)
+        {
+                exit(EXIT_FAILURE);
```

Comment:
replace with something like LOG_FATAL.

Also I note some style inconsistency. Sometimes the { is after the if and sometimes it's on the next line. We should follow whatever the neper style is and be consistent about it.